pkg/controller/datavolume-controller.go

```diff
@@ -2601,3 +2613,59 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
 func newLongTermCloneTokenGenerator(key *rsa.PrivateKey) token.Generator {
 	return token.NewGenerator(common.ExtendedCloneTokenIssuer, key, 10*365*24*time.Hour)
 }
+
+func (r *DatavolumeReconciler) garbageCollect(dataVolume *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) (*reconcile.Result, error) {
+	if dataVolume.Status.Phase != cdiv1.Succeeded {
+		return nil, nil
+	}
+	cdiConfig := &cdiv1.CDIConfig{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: common.ConfigName}, cdiConfig); err != nil {
+		return nil, err
+	}
+	dvTTL := cdiConfig.Spec.DataVolumeTTLSeconds
+	if dvTTL == nil || *dvTTL < 0 {
+		return nil, nil
+	}
+	// Current DV still has TTL, so reconcile will return with the needed RequeueAfter
+	if delta := getDeltaTTL(dataVolume, *dvTTL); delta > 0 {
+		return &reconcile.Result{RequeueAfter: delta}, nil
+	}
+	if err := r.detachPvcDeleteDv(pvc, dataVolume); err != nil {
+		return nil, err
+	}
+	return &reconcile.Result{}, nil
+}
+
+func (r *DatavolumeReconciler) detachPvcDeleteDv(pvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume) error {
+	dvDelete := dv.Annotations[AnnDeleteAfterCompletion]
+	if IsSucceeded(pvc) && dvDelete == "true" {
+		updatePvcOwnerRefs(pvc, dv)
+		delete(pvc.Annotations, AnnPopulatedFor)
+		if err := r.updatePVC(pvc); err != nil {
+			return err
+		}
+		if err := r.client.Delete(context.TODO(), dv); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func updatePvcOwnerRefs(pvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume) {
```

Comment:
I would call this function something else like removeDVasOwnerFrom and make the first argument an Object this way we can potentially re-use it to remove the owner from other objects as well.

Reply:
yep, much better name

Reply:
Right, good point.